### ISSUE TITLE
Private `#[no_mangle]` symbols are exported from a `cdylib`

### ISSUE THREAD

**Issue by nxtn-staged** [rust-lang/rust#98449]
In C++ (MSVC) un-mangling and exporting are controlled via `extern "C"` and `__declspec(dllexport)` respectively.

```cpp
extern "C" {
    namespace to_export {
        __declspec(dllexport) int64_t foo() {
            int64_t bar();
            return bar();
        }
    }
    namespace not_to_export {
        int32_t bar() {
            return 42;
        }
    }
}
```

```
    00000000 characteristics
    FFFFFFFF time date stamp
        0.00 version
           1 ordinal base
           1 number of functions
           1 number of names

    ordinal hint RVA      name

          1    0 00001000 foo = foo
```

In Rust, however, all `#[no_mangle]` functions are exported from the final `cdylib`.

```rust
pub mod to_export {
    #[no_mangle]
    pub unsafe extern "system" fn foo() -> i64 {
        extern "system" {
            fn bar() -> i64;
        }
        bar()
    }
}
mod not_to_export {
    #[no_mangle]
    extern "system" fn bar() -> i32 {
        42
    }
}
```

```
    00000000 characteristics
    FFFFFFFF time date stamp
        0.00 version
           1 ordinal base
           2 number of functions
           2 number of names

    ordinal hint RVA      name

          1    0 00001000 bar = bar
          2    1 00001010 foo = foo
```

It seems that it is not possible to define an un-mangled and un-exported symbol in Rust.

**Comment by dvtkrlbs** [comment]
I would like to try to tackle this. However I need some guidance for a starting point.

**Comment by SkiFire13** [comment]
My guess is that this is related to https://github.com/rust-lang/rust/pull/95604

**Comment by ehuss** [comment]
I believe exporting the symbol is intentional as documented [here](https://doc.rust-lang.org/nightly/reference/abi.html#the-no_mangle-attribute). There is also some more information at #54451.  Can you say more about why you would want an unmangled symbol that isn't exported?

**Comment by nxtn-staged** [comment]
When building `#![no_std]` crates or linking with static libraries, I'm able to fix `unresolved external symbol` errors with `#[no_mangle]` symbols, but I do not expect them to appear in the export list.

**Comment by ChrisDenton** [comment]
Yes, in general there is a difference between making a symbol for the linker to find and exporting it from the final binary. According to the [`no_mangle` reference](https://doc.rust-lang.org/reference/abi.html#the-no_mangle-attribute):

> Additionally, the item will be publicly exported from the produced library or object file, similar to the [`used` attribute](https://doc.rust-lang.org/reference/abi.html#the-used-attribute).

And the `used` reference says (my bold):

> This [attribute](https://doc.rust-lang.org/reference/attributes.html) forces the compiler to keep the variable in the output object file (.o, .rlib, etc. **excluding final binaries**) even if the variable is not used, or referenced, by any other item in the crate. However, **the linker is still free to remove such an item**.
